politeiad/backendv2/tstorebe/plugins/pi/cmds.go

```diff
@@ -242,25 +242,131 @@ func (p *piPlugin) cmdBillingStatusChanges(token []byte) (string, error) {
 	return string(reply), nil
 }
 
+// cacheDataGet retrieves the data associated with the given token from the
+// memory cache.  If data doesn't exist in cache it returns nil.
+func (p *piPlugin) cacheDataGet(token string) *cacheData {
+	p.cache.Lock()
+	defer p.cache.Unlock()
+
+	return p.cache.data[token]
+}
+
+// cacheVoteStatusSet stores the vote status associated with the given token
+// in cache. If the cache is full and a new entry is being added, the oldest
+// is removed from the cache.
+func (p *piPlugin) cacheVoteStatusSet(token string, status ticketvote.VoteStatusT) {
+	p.cache.Lock()
+	defer p.cache.Unlock()
+
+	// If an entry associated with the proposal already exists in cache
+	// overwrite the vote status.
+	if p.cache.data[token] != nil {
+		p.cache.data[token].voteStatus = &status
+		return
+	}
+
+	// If entry does not exist and cache is fulli, then remove oldest entry
+	if p.cache.entries.Len() == piCacheLimit {
+		// Remove front - oldest entry from entries list.
+		t := p.cache.entries.Remove(p.cache.entries.Front()).(string)
+		// Remove oldest status from map.
+		delete(p.cache.data, t)
+	}
+
+	// Store new status.
+	p.cache.entries.PushBack(token)
+	p.cache.data[token] = &cacheData{
+		voteStatus: &status,
+	}
+}
+
+// cacheProposalStatusSet stores the proposal status associated with
+// the given token in cache. If the cache is full and a new entry is being
+// added, the oldest entry is removed from the cache.
+func (p *piPlugin) cacheProposalStatusSet(token string, status pi.PropStatusT) {
+	p.cache.Lock()
+	defer p.cache.Unlock()
+
+	// If an entry associated with the proposal already exists in cache
+	// overwrite the proposal status.
+	if p.cache.data[token] != nil {
+		p.cache.data[token].proposalStatus = &status
+		return
+	}
+
+	// If entry does not exist and cache is fulli, then remove oldest entry
+	if p.cache.entries.Len() == piCacheLimit {
+		// Remove front - oldest entry from entries list.
+		t := p.cache.entries.Remove(p.cache.entries.Front()).(string)
+		// Remove oldest status from map.
+		delete(p.cache.data, t)
+	}
+
+	// Store new status.
+	p.cache.entries.PushBack(token)
+	p.cache.data[token] = &cacheData{
+		proposalStatus: &status,
+	}
+}
+
 // cmdSummary returns the pi summary of a proposal.
 func (p *piPlugin) cmdSummary(token []byte) (string, error) {
```

Comment:
The logic in this function needs to be cleaned up and made easier to read.  I would expect it to read something like this.  This is just an example, you don't need to follow it exactly, but it demonstrates what I mean by easier to read.

```
	// Check if the proposal status has been cached
	cachedStatus := p.proposalStatuses.get(token)
	if statusIsFinal(cachedStatus) {
		// Status is final cannot be changed. No
		// need to fetch any additional data.
		goto reply
	}
	if statusRequiresRecord(cachedStatus) {
		// Fetch the record data
	}
	if statusRequiresVoteSummary(cachedStatus) {
		// Fetch the vote summary
	}
	if statusRequiresBillingStatuses(cachedStatus) {
		// Fetch the billing statuses
	}

	// Determine the proposal status
	s, err = proposalStatus(mdState, mdStatus, voteStatus, voteMD, bscs)
	if err != nil {
		return "", err
	}

	// Cache the vote status result
	p.proposalStatuses.set(token, voteMD, s)
```